### DISCUSSION TITLE
Expected Frame pending field from radio driver for Outcoming ACK ?

### DISCUSSION THREAD

**Discussion by testter2** [openthread/openthread#8677] [Q&A] [answered]
Hi All,

I'm just asking about the expected frame pending bit in the outcoming ACK for different types of frames.

As per IEEE Std 802.15.4-2015, I can see that it states that frame pending bit for outcoming ACK is determined whether the device has a frame pending or not (if coordinator has enough time to determine it ) and this is happening only for **data request command** .

My question is related to what is expected from radio driver for frame pending bit in ACK:
should it handled for any ACK regardless the frame type or for mac  data request command only ?

Thanks in advance.

**Comment by abtink** [comment, marked as answer]
@testter2 for Thread 1.2 onwards, the radio platform must set the frame-pending accordingly on acks to all frame types (not just data request command).

----

In `mac.cpp`, method `Mac::HandleTransmitDone()`:
https://github.com/openthread/openthread/blob/30aebb5e74064d63513ea78dd4fa2ecc46fd38f9/src/core/mac/mac.cpp#L1405-L1407
And then `DataPollSender` checks the `aAckFrame` to decide whther to poll or not.
https://github.com/openthread/openthread/blob/30aebb5e74064d63513ea78dd4fa2ecc46fd38f9/src/core/mac/data_poll_sender.cpp#L358-L370

**Comment by testter2** [comment]
Thanks for quick answer.
Is this follow IEEE Std 802.15.4 ? and Is there any issue explains the reason behind this ?

**Reply by abtink** [reply]
This behavior is part Thread 1.2 specification and specified in the section 3.2.6.2 titled "Enhanced Frame Pending". It describes the justification for this as well.

In short,  any data frame sent from a sleepy child can also act as a data poll which help it avoid sending extra data polls and thus save energy (battery life).

**Reply by testter2** [reply]
Thanks, it's clear now.
